README.md
# Typeahead

Run these commands from the command-line.
#### Dependencies
```
npm install
bower install
```

#### Build
```
gulp default
```

#### Unit Tests
```
karma start --single-run
```

## Options
Name | Description
--- | ---
appendToBody | If `true`, the dropdown-menu will be appended to the HTML body element, instead of inside the autocomplete container. Default: `false`
dataSource | Can be either an Array or an URL. In case of an URL, Ajax is used to gather the data.
nameProperty | The property that will be used as display text. Default: `name`
valueProperty | The property that will be used as the value. Default: `value`
selectableProperty | The property that will be used to specify if an element can be selected. If an element has `selectable = false` it cannot be clicked/selected. Default: `selectable`
notSelectableClass | The class that will be rendered on the dropdown item, if it can not be selected. Default: `no-select`
valueField | Can be either a selector or an element. Specified elements will get their value set to the value of the selected element (specified by the `valueProperty` option).
filter | `function(input, data) { ... }` Function which will be used to filter the items. `input` contains the input value. `data` contains the object retrieved by the dataSource.
filterOn | Event which will trigger the filter function. Will be attached to the text input field. Default: `input`
validation | `function(input, data) { ... }` Function which can be used for custom validations. Has to return `true` if value is valid, otherwise the input field will be marked as invalid.
validateOn | Event which will trigger the validation function. Will be attached to the text input field. Default: `blur`
onSelected | Function that will be called when a value is selected.
invalidClass | Class that will be set on the input, if the validation function does not return `true`. Default: `invalid`

## Events
Name | Description
--- | ---
initial-value-selected | Will be triggered as soon as the initial value is selected. Initial value is optionally provided by the referenced `valueField`. The value will only be selected if the `dataSource` contains a corresponding element.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.2] - 22.03.2018
### Added
- Render class on dropdown item if it is not selectable.

## [1.0.1] - 21.03.2018
### Added
- selectableProperty property and underlying functionality. Elements can be marked as not selectable.

## [1.0.0] - 28.02.2018
### Added
- onSelected property
- Changelog

### Changed
- Updated license

test/typeahead.spec.js
'use strict';

describe('Typeahead', function () {
    var $ = jQuery;

    jasmine.getFixtures().fixturesPath = 'base/test/fixtures/';

    const fireInputEvent = (element) => {
        fireEvent(element, 'input');    // fire the input event as if someone was typing
    };

    const fireBlurEvent = (element) => {
        fireEvent(element, 'blur');
    };

    const fireEvent = (element, eventName) => {
        var evt = document.createEvent("HTMLEvents");
        evt.initEvent(eventName, false, true);
        element.dispatchEvent(evt);
    };

    const keyPress = (element, key) => {
        if (!element) {
            element = document;
        }
        var event = document.createEvent('Event');
        event.keyCode = key;
        event.initEvent('keydown');
        element.dispatchEvent(event);
    };

    const hiddenFieldSelector = '#test-hidden-field';
    const TEST_VALUE = 'X';
    const TEST_METHOD = jasmine.createSpy('getMethod spy').and.callFake((url, {value}) => {
        return new Promise(function (resolve) {
            var data = [];
            for (var i = 0; i < 15; i++) {
                data.push({name: value + i, value: i})
            }
            resolve(data);
        });
    });

    // inject the HTML fixture for the tests
    beforeEach(function () {
        loadFixtures('typeahead-fixture.html');
    });

    // remove the html fixture from the DOM
    afterEach(function () {
        $('#fixture').remove();
    });

    it('should create DOM elements', function () {
        $('.test-element').typeahead();

        var typeahead = $('.typeahead');
        expect(typeahead).toBeInDOM();
        expect($('ul.items.dropdown-menu')).toBeInDOM();
    });

    it('should call the url provided as data attribute on input', function () {
        const element = $('.test-element');
        const TEST_URL = 'TEST_URL';
        const spy = jasmine.createSpy("callback spy").and.callFake((url) => {
            return new Promise((resolve) => {
                expect(url).toBe(TEST_URL);
                resolve();
            });
        });

        element.attr('data-source', TEST_URL);
        element.typeahead({getMethod: spy});

        const typeahead = $('.typeahead');

        expect(spy).not.toHaveBeenCalled();
        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);
        expect(spy).toHaveBeenCalled();
    });

    it('should only open the dropdown-menu if it is not empty', function (done) {
        const element = $('.test-element');
        element.typeahead({getMethod: TEST_METHOD});

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        expect(dropdown).toBeHidden();

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            expect(dropdown).toBeVisible();
            done();
        }, 100);

    });

    it('should select dropdown item on click and hide the menu', function (done) {
        const element = $('.test-element');
        element.typeahead({getMethod: TEST_METHOD});

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            expect(dropdown).toBeVisible();
            dropdown.find('li:eq(2)').click();
            expect(TEST_METHOD).toHaveBeenCalled();

            expect(typeahead[0].getAttribute('selected')).toBe('2');
            expect(typeahead).toHaveValue(TEST_VALUE + '2');
            expect(dropdown).toBeHidden();

            done();
        }, 200);
    });

    it('should set the value of the provided hidden field', function (done) {
        $('.test-element').typeahead({getMethod: TEST_METHOD, valueField: '#test-hidden-field'});

        let hiddenField = $('#test-hidden-field');
        expect(hiddenField).toHaveValue('');

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            dropdown.find('li:eq(2)').click();
            expect(hiddenField).toHaveValue('2');
            done();
        }, 200);
    });

    it('should set the value of the field provided as data attribute', function (done) {
        $('.test-element').attr('data-value-field', hiddenFieldSelector);
        $('.test-element').typeahead({getMethod: TEST_METHOD});

        let hiddenField = $(hiddenFieldSelector);
        expect(hiddenField).toHaveValue('');

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            dropdown.find('li:eq(4)').click();
            expect(hiddenField).toHaveValue('4');
            done();
        }, 200);
    });


    it('should clear the selected value from the attribute and the hidden field if selection is cleared', function (done) {
        const $input = $('.test-element');
        $input.typeahead({getMethod: TEST_METHOD, valueField: hiddenFieldSelector});

        const typeahead = $('.typeahead');
        const hiddenField = $('#test-hidden-field');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            $('.typeahead ~ ul.items.dropdown-menu > li:eq(2)').click();

            // test if initial state of this test is correct
            expect(hiddenField).toHaveValue('2');
            expect(+$('.typeahead')[0].getAttribute('selected')).toBe(2);
            expect($input).toHaveValue(TEST_VALUE + '2');

            $input.val('');
            fireInputEvent($input[0]);  // fire the input event as if someone was typing

            setTimeout(() => {
                expect($('.typeahead')[0].getAttribute('selected')).toBe('');
                expect(hiddenField).toHaveValue('');
                done();
            }, 200);
        }, 100);
    });

    it('should select the next element if the down key is pressed', function (done) {
        const element = $('.test-element');
        element.typeahead({getMethod: TEST_METHOD});

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        typeahead.focus();

        setTimeout(() => {
            expect(dropdown).toBeVisible();
            expect($('.typeahead ~ ul.items.dropdown-menu > li:eq(1)')).not.toBeFocused();

            // test setup done. test starts here
            keyPress(typeahead[0], 40);   // down arrow
            expect($(document.activeElement)).toEqual('.typeahead ~ ul.items.dropdown-menu > li:eq(0)');
            expect($('.typeahead ~ ul.items.dropdown-menu > li:eq(0)')).toBeFocused();
            expect($(document.activeElement)).toContainText(TEST_VALUE + '0');

            keyPress(dropdown[0], 40);   // down arrow
            expect($(document.activeElement)).toEqual('.typeahead ~ ul.items.dropdown-menu > li:eq(1)');
            expect($('.typeahead ~ ul.items.dropdown-menu > li:eq(1)')).toBeFocused();
            expect($(document.activeElement)).toContainText(TEST_VALUE + '1');

            keyPress(dropdown[0], 38);   // up arrow
            expect($(document.activeElement)).toEqual('.typeahead ~ ul.items.dropdown-menu > li:eq(0)');
            expect($('.typeahead ~ ul.items.dropdown-menu > li:eq(0)')).toBeFocused();
            expect($(document.activeElement)).toContainText(TEST_VALUE + '0');

            done();
        }, 200);
    });

    it('should close the dropdown menu on ESC-key', function (done) {
        const element = $('.test-element');
        element.typeahead({getMethod: TEST_METHOD});

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            expect(dropdown).toBeVisible();
            // test setup done. test starts here
            keyPress(document, 27);   // esc key
            expect(dropdown).toBeHidden();
            done();
        }, 100);
    });

    it('should clear the dropdown-items if the api returns an empty or invalid result', function (done) {
        const numberOfItems = 15;
        const EMPTY_RETURN_VALUE = 'RETURN_EMPTY';
        const INVALID_RETURN_VALUE = 'RETURN_INVALID';
        const spy = jasmine.createSpy('getMethod spy').and.callFake((url, {value}) => {
            return new Promise(function (resolve) {
                if (value === INVALID_RETURN_VALUE) {
                    resolve("asdfasdfasdf");
                }

                let data = [];
                if (value !== EMPTY_RETURN_VALUE) {
                    for (let i = 0; i < numberOfItems; i++) {
                        data.push({name: value + i, value: i})
                    }
                }
                resolve(data);
            });
        });

        const element = $('.test-element');
        element.typeahead({getMethod: spy});

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            expect(spy).toHaveBeenCalledTimes(1);
            expect(dropdown).toBeVisible();
            expect(dropdown.children()).toHaveLength(numberOfItems);

            // test setup done. test starts here
            typeahead.val(EMPTY_RETURN_VALUE);
            fireInputEvent(typeahead[0]);

            setTimeout(() => {
                // Empty array returned. Should have no items in dropdown
                expect(spy).toHaveBeenCalledTimes(2);
                expect(dropdown).toBeVisible();
                expect(dropdown.children()).not.toHaveLength(numberOfItems);
                expect(dropdown.children()).toHaveLength(0);

                // Reset to valid searchstring. Should have 15 items in dropdown
                typeahead.val(TEST_VALUE);
                fireInputEvent(typeahead[0]);

                setTimeout(() => {
                    expect(spy).toHaveBeenCalledTimes(3);
                    expect(dropdown).toBeVisible();
                    expect(dropdown.children()).toHaveLength(numberOfItems);

                    // INvalid return value. Should not crash and have no items in dropdown
                    typeahead.val(INVALID_RETURN_VALUE);
                    fireInputEvent(typeahead[0]);

                    setTimeout(() => {
                        expect(spy).toHaveBeenCalledTimes(4);
                        expect(dropdown).toBeVisible();
                        expect(dropdown.children()).not.toHaveLength(numberOfItems);
                        expect(dropdown.children()).toHaveLength(0);

                        done();
                    }, 100);
                }, 100);
            }, 100);
        }, 100);
    });

    it('should call onSelected function', function () {
        const selectedFunc = jasmine.createSpy('onSelected spy');
        $('.test-element').typeahead({getMethod: TEST_METHOD, onSelected: selectedFunc});

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        dropdown.find('li:eq(2)').click();
        expect(selectedFunc).toHaveBeenCalled();
    });

    it('should call onSelected function with itself as parameter', function () {
        const selectedFunc = jasmine.createSpy('onSelected spy');
        $('.test-element').typeahead({getMethod: TEST_METHOD, onSelected: selectedFunc});

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        dropdown.find('li:eq(2)').click();
        expect(selectedFunc).toHaveBeenCalledWith(typeahead.data('typeahead'));
    });

    it('should not select dropdown item that is not selectable and should not hide the menu', function (done) {
        const element = $('.test-element');
        const NOT_SELECTABLE_TEXT = "not selectable";
        const NOT_SELECTABLE_VALUE = 999;
        element.typeahead({
            getMethod: jasmine.createSpy('getMethod spy').and.callFake((url, {value}) => {
                return new Promise(function (resolve) {
                    resolve([{name: NOT_SELECTABLE_TEXT, value: NOT_SELECTABLE_VALUE, selectable: false}]);
                });
            })
        });

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            expect(TEST_METHOD).toHaveBeenCalled();

            expect(dropdown).toBeVisible();
            dropdown.find('li').click();

            expect(typeahead[0].getAttribute('selected')).not.toBe(NOT_SELECTABLE_VALUE);
            expect(typeahead).not.toHaveValue(NOT_SELECTABLE_VALUE);
            expect(dropdown).not.toBeHidden();

            done();
        }, 200);
    });

    it('should render class on not selectable items', function (done) {
        const element = $('.test-element');
        const NOT_SELECTABLE_TEXT = "not selectable";
        const NOT_SELECTABLE_CLASS = "IS-NOT-SELECTABLE";
        const NOT_SELECTABLE_VALUE = 999;
        element.typeahead({
            getMethod: jasmine.createSpy('getMethod spy').and.callFake((url, {value}) => {
                return new Promise(function (resolve) {
                    resolve([{name: NOT_SELECTABLE_TEXT, value: NOT_SELECTABLE_VALUE, selectable: false}]);
                });
            }),
            notSelectableClass: NOT_SELECTABLE_CLASS
        });

        const typeahead = $('.typeahead');
        let dropdown = $('.typeahead ~ ul.items.dropdown-menu');

        typeahead.val(TEST_VALUE);
        fireInputEvent(typeahead[0]);

        setTimeout(() => {
            expect(TEST_METHOD).toHaveBeenCalled();

            expect(dropdown).toBeVisible();
            expect(dropdown.find('li')).toHaveClass(NOT_SELECTABLE_CLASS);

            done();
        }, 200);
    });

    /*
     it('should only append the dropdown-menu items if the dropdown is open', function () {
     let options = {dataSource: test_dataSource};
     $('.test-element').autocomplete(options);

     let dropdown = $('.autocomplete > ul.items.dropdown-menu');

     expect(dropdown).toBeInDOM();
     expect(dropdown).toBeEmpty();

     let button = $('.autocomplete button.btn.btn-default');
     button.click();

     expect(dropdown).not.toBeEmpty();

     expect(dropdown.children().length).toBe(3);
     expect(dropdown.children().eq(1).text()).toBe(options.dataSource[1].name);
     expect(+dropdown.children().eq(2).attr('value')).toBe(options.dataSource[2].value);

     button.click();
     expect(dropdown).toBeEmpty();
     });

     it('should open the dropdown-menu on button click', function () {
     let options = {dataSource: test_dataSource};
     $('.test-element').autocomplete(options);

     let dropdown = $('.autocomplete > ul.items.dropdown-menu');

     expect(dropdown).toBeHidden();
     $('.autocomplete button.btn.btn-default').click();
     expect(dropdown).toBeVisible();
     });


     it('should initially select the corresponding value of the hidden field', function () {
     let options = {dataSource: test_dataSource, valueField: hiddenFieldSelector};
     let hiddenField = $('#test-hidden-field');
     hiddenField.val(test_dataSource[1].value);

     $('.test-element').autocomplete(options);

     let container = $('.autocomplete.input-group');

     expect(hiddenField).toHaveValue('' + test_dataSource[1].value);
     expect($('.test-element')[0].value).toBe('' + test_dataSource[1].name);
     expect($('.autocomplete')[0].getAttribute('selected')).toBe('' + test_dataSource[1].value);
     });


     it('should by default filter on the input event', function () {
     let $input = $('.test-element');
     let filterFunc = jasmine.createSpy("filter() spy");
     let options = {
     dataSource: test_dataSource,
     filter: filterFunc
     };

     $input.autocomplete(options);
     $input.val(test_dataSource[1].name);

     // fire the events
     fireBlurEvent($input[0]);
     expect(filterFunc).not.toHaveBeenCalled();

     fireInputEvent($input[0]);
     expect(filterFunc).toHaveBeenCalled();
     });*/
});

gulpfile.js
'use strict';
const gulp = require('gulp'),
    uglify = require('gulp-uglify'),
    rename = require('gulp-rename'),
    babel = require('gulp-babel'),
    concat = require('gulp-concat'),
    server = require('karma').Server,
    wrap = require("gulp-wrap");

// Concatenate & Minify JS
gulp.task('scripts', function () {
    return gulp.src('src/*.js')
        .pipe(concat('typeahead.js'))
        .pipe(wrap('(function ($) {<%= contents %>}(window.jQuery));'))
        .pipe(babel({
            presets: ['es2015']
        }))
        .pipe(gulp.dest('dist'))
        .pipe(rename('typeahead.min.js'))
        .pipe(uglify())
        .pipe(gulp.dest('dist'));
});

gulp.task('test', ['scripts'], function (done) {
    new server({
        configFile: __dirname + '/karma.conf.js',
        singleRun: true
    }, function (exitCode) {
        done();
        process.exit(exitCode);
    }).start();
});

// Watch for file changes and re-run tests on each change
gulp.task('tdd', function (done) {
    new server({
        configFile: __dirname + '/karma.conf.js'
    }, function () {
        done();
    }).start();
    return gulp.watch('src/*.js', ['scripts']);
});

gulp.task('deploy', ['scripts']);
gulp.task('default', ['scripts', 'test']);


karma.conf.js
// Karma configuration
// Generated on Wed Jan 11 2017 12:26:58 GMT+0100 (Mitteleuropäische Zeit)

module.exports = function (config) {
    config.set({

        // base path that will be used to resolve all patterns (eg. files, exclude)
        basePath: '',


        // frameworks to use
        // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
        frameworks: ['jasmine-jquery', 'jasmine'],


        // list of files / patterns to load in the browser
        files: [
            'dist/typeahead.js',
            'test/**/*.spec.js',
            { pattern: 'test/fixtures/*.html', included: false, served: true, watched: true }
        ],


        // list of files to exclude
        exclude: [
            '*.min.js'
        ],


        // preprocess matching files before serving them to the browser
        // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
        preprocessors: {},


        // test results reporter to use
        // possible values: 'dots', 'progress'
        // available reporters: https://npmjs.org/browse/keyword/karma-reporter
        reporters: ['progress'],


        // web server port
        port: 9876,


        // enable / disable colors in the output (reporters and logs)
        colors: true,


        // level of logging
        // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        logLevel: config.LOG_INFO,

        browserConsoleLogOptions: {
            terminal: true,
            level: ""
        },

        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: true,


        // start these browsers
        // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
        browsers: ['Chrome'],


        // Continuous Integration mode
        // if true, Karma captures browsers, runs the tests and exits
        singleRun: false,

        // Concurrency level
        // how many browser should be started simultaneous
        concurrency: Infinity
    })
}


src/typeahead.js
'use strict';

class Typeahead {

    constructor(element, options) {
        this.options = options;
        if (this.options.filter) {
            this.options.filter = this.options.filter.bind(this);
        }

        this.$input = $(element);
        this.$input.addClass('typeahead');
        this.$input.attr('autocomplete', 'off');    // autocomplete="off" to prevent default browser menus

        if (this.$input.data('value-field')) {
            this.$valueField = $(this.$input.data('value-field'));
        } else if (typeof this.options.valueField === "string") {
            this.$valueField = $(this.options.valueField);
        } else if (this.options.valueField && this.options.valueField.length) {
            this.$valueField = this.options.valueField;
        }

        this.datasource = this.$input.data('source');
        if (!this.datasource) {
            this.datasource = this.options.dataSource;
        }
        this.data = null;

        this.$items = $('<ul class="items dropdown-menu typeahead-menu" style="display:none;" role="menu"></ul>');
        this.$items.insertAfter(this.$input);

        if (this.options.appendToBody) {
            let id = '' + (10000 * Math.random() * new Date().getTime() * window.outerHeight);
            this.$items.attr('id', id);
            this.$input.attr('dd-menu', id);
            this.$items.detach().appendTo('body');
        }

        // initialize event listeners
        let _this = this;
        this.$input[0].addEventListener(this.options.searchOn, function () {
            _this.search(this.value);
        });

        this.$input.click(() => {
            if (this.data && this.data.length) {
                this.open = true;
            }
        });

        // handle arrow keys
        this.$input.on('keydown', function (e) {
            if (e.keyCode === 40) {
                _this.$items.children().first().focus();
            }
        });

        this.$items.on('keydown', (e) => {
            if (e.keyCode === 40) {
                $(document.activeElement).next().focus();   // DOWN
            } else if (e.keyCode === 38) {
                $(document.activeElement).prev().focus();   // UP
            } else if (e.keyCode === 9) {
                e.preventDefault(); // TAB
            }
        });

        // close the dropdown menu if clicked outside or on esc keypress
        $('body').click(() => {
            const $active = $(document.activeElement);
            if (!$active.is(this.$input) && !$active.is(this.$items) && !this.$items.find($active).length) {
                this.open = false
            }
        });
        document.addEventListener('keydown', (e) => {
            if (e.keyCode === 27) {
                this.open = false;  // esc key
            }
        });

        this.open = false;
    }

    get open() {
        return !!this.$input.attr('open');
    }

    set open(val) {
        // reflect the value of the open property as an HTML attribute
        if (val) {
            this.$input.attr('open', '');
        } else {
            this.$input.removeAttr('open');
        }

        if (this.open) {
            this.buildDropdownItems(this.data);
            this.$items.show();
            this.setMenuDirection();
        } else {
            this.$items.hide();
            this.destroyDropdownItems();
        }
    }

    get selected() {
        return this.$input.data('selected');
    }

    set selected(value) {
        // reflect the value of the selected property as an HTML attribute
        if (!value) {
            value = {};
        }

        if (typeof value !== "object" && this.data) {
            if (!isNaN(+value)) {
                value = +value; // if value is a string we try to convert it to a number, otherwise we leave it as a string
            }
            let elem = this.data.filter(x => x[this.options.valueProperty] === value);
            if (elem && elem.length && elem[0]) {
                value = elem[0];
            }
        }

        this.$input[0].setAttribute('selected', value[this.options.valueProperty] || '');
        this.$input.data('selected', value);

        if (this.$valueField) {
            this.$valueField.val(value[this.options.valueProperty] || '');
        }

        if (this.options.onSelected) {
            this.options.onSelected(this);
        }

        this.$input.trigger('change');
    }

    search(input) {
        if (!input) {
            input = '';
        }
        this.selected = null;

        let results;
        this.options.getMethod(this.datasource, {value: input}).then((data) => {
            this.data = data;
            this.buildDropdownItems(results);
            this.open = true;
        });
    }

    buildDropdownItems(dataItems) {
        this.destroyDropdownItems();

        if (!dataItems || !dataItems.length)
            return;

        let _this = this;
        dataItems.forEach(x => {
            let li = document.createElement('li');
            li.setAttribute('value', x[_this.options.valueProperty]);
            li.setAttribute('tabindex', '-1');
            li.innerHTML = '<a href="#">' + x[_this.options.nameProperty] + '</a>';
            const selectItemHandler = (e) => {
                _this.$input.val(x[_this.options.nameProperty]);
                _this.selected = x;
                _this.open = false;
            };

            if (x[_this.options.selectableProperty] === undefined || x[_this.options.selectableProperty] === null || !!x[_this.options.selectableProperty] === true) {
                li.addEventListener('click', selectItemHandler);
                li.addEventListener('keydown', (e) => {
                    if (e.keyCode === 13) {
                        // enter key
                        selectItemHandler(e);
                        _this.$input.focus();
                        e.preventDefault();
                    }
                });
            } else if (!!x[_this.options.selectableProperty] === false) {
                li.addEventListener('click', () => _this.open = true);
                li.setAttribute('class', _this.options.notSelectableClass);
            }
            _this.$items.append(li);
        });
    }

    destroyDropdownItems() {
        this.$items.children().remove();
    }

    toggleOpen() {
        this.open = !this.open;
    }

    setMenuDirection() {
        let inputOffset = this.$input.offset();
        let inputHeight = this.$input.outerHeight();
        let inputMarginTop = parseInt(this.$input.css('margin-top'));

        // let menuOffset = this.$items.offset();
        let menuHeight = this.$items.outerHeight();

        let vpHeight = $(window).height();

        let noSpaceBelow = inputOffset.top + inputHeight + menuHeight > vpHeight;
        let spaceAbove = inputOffset.top - $(window).scrollTop() - menuHeight > 0;

        if (noSpaceBelow && spaceAbove) {
            this.$items.offset({top: inputOffset.top - menuHeight - inputMarginTop, left: inputOffset.left})
        } else {
            this.$items.offset({top: inputOffset.top + inputHeight + inputMarginTop, left: inputOffset.left})
        }
    }
}

const defaultOptions = {
    nameProperty: 'name',
    valueProperty: 'value',
    selectableProperty: 'selectable',
    notSelectableClass: 'no-select',
    valueField: null,
    dataSource: null,
    searchOn: 'input',
    appendToBody: false,
    getMethod: $.get,
    onSelected: null
};

$.fn.typeahead = function (option) {
    this.each(function () {
        let $this = $(this),
            data = $this.data('typeahead'),
            options = typeof option === 'object' && option;
        if (!data) {
            let opts = $.extend({}, defaultOptions, options);
            let typeahead = new Typeahead(this, opts);
            $this.data('typeahead', typeahead);
        }
    });
    return this;
};

$(() => {
    $('[data-provide="softec-typeahead"]').typeahead();
});